editor/utils/heartbeat-post-locking.js

```diff
@@ -0,0 +1,83 @@
+/* global ajaxurl */
+
+/**
+ * WordPress dependencies
+ */
+import { dispatch, select } from '@wordpress/data';
+
+export function setupHearthbeatPostLocking() {
+	/**
+	 * Configure Heartbeat post locks.
+	 *
+	 * Used to lock editing of an object by only one user at a time.
+	 *
+	 * When the user does not send a heartbeat in a heartbeat-tick
+	 * the user is no longer editing and another user can start editing.
+	 */
+	jQuery( document ).on( 'heartbeat-send.refresh-lock', function( e, data ) {
+		const lock = jQuery( '#active_post_lock' ).val(),
+			postId = jQuery( '#post_ID' ).val(),
+			send = {};
+
+		if ( ! postId ) {
+			return;
+		}
+
+		// Check if the post is already locked by another user.
+		const locked = select( 'core/editor' ).isPostLocked();
+		if ( locked ) {
+			return;
+		}
+
+		send.post_id = postId;
+
+		if ( lock ) {
+			send.lock = lock;
+		}
+
+		data[ 'wp-refresh-post-lock' ] = send;
+	} )
+
+		// Refresh post locks: update the lock string or show the dialog if somebody has taken over editing.
+		.on( 'heartbeat-tick.refresh-lock', function( e, data ) {
+			if ( data[ 'wp-refresh-post-lock' ] ) {
+				const received = data[ 'wp-refresh-post-lock' ];
+				if ( received.lock_error ) {
+					// Auto save and display the takeover modal.
+					dispatch( 'core/editor' ).autosave();
```

Comment:
What's the reasoning here? Autosave to create a revision in case of content conflicts?

Reply:
this happens when another user takes over the post you are editing. you may have unsaved changes and once another user takes over you are locked out of editing. Firing an autosave insures the user's edits aren't los.

This  matches what happens currently in the classic editor. To see this behavior (in the classic editor), edit a post, make some changes but don't update, then, open the post as another user and take over the post. the original user gets locked out, and the modal shows a spinner and autosave confirmation.